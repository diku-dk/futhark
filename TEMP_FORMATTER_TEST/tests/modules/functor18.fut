-- A functor whose parameter is itself a functor.
-- ==
-- input { 1 } output { 3 4 7 }

module F = \(P_f: (P_f_a: {type a val f: a -> a}) -> {val f: P_f_a.a -> P_f_a.a}) ->
           \(P_x: {type a val f: a -> a}) ->
            {type a = P_x.a open (P_f P_x)}

module twice = \(twice_P: {type a val f: a -> a}) -> {
  def f (x: twice_P.a) = twice_P.f (twice_P.f x)
}

module thrice = \(thrice_P: {type a val f: a -> a}) -> {
  def f (x: thrice_P.a) = thrice_P.f (thrice_P.f (thrice_P.f x))
}

module add_one = {type a = i32 def f(x: i32) = x + 1}
module F_2 = F twice add_one
module F_3 = F thrice add_one
module F_6 = F twice F_3

def main(x: i32) = (F_2.f x, F_3.f x, F_6.f x)
