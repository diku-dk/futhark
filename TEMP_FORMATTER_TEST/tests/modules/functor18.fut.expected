955b43d3334d96561781eab2050182f3
